### PR TITLE
comments and urls in tasks exception warning

### PR DESCRIPTION
**Pull request by peter-ochieng**
https://github.com/medic/cht-core/issues/6506

Throw exception in medic conf during upload when tasks.js contains comments or urls.

### CODE REVIEW

**Comment by peter-ochieng** [comment]
@garethbowen Could you advise on what am missing on the failing test and if I've applied the changes as you described?

**Comment by peter-ochieng** [comment]
@garethbowen could you have another look?

File: test/lib/compile-nools-rules.spec.js [hunk 1]
```diff
@@ -103,4 +103,19 @@ describe('compile nools-rules', () => {
...
       });
   });
+
+  it('comments and urls in fields yields exception', () => {
+    const mocks = genMocks();
+    mocks.fs.exists
+      .withArgs('/rules.nools.js').returns(false)
+      .withArgs('/targets.js').returns(true)
+      .withArgs('/tasks.js').returns(true);
+    mocks.fs.read.withArgs('/rules.nools.js').returns('define Task {_id: null, title: "http://foo.bar.com"}');
```
Comment (garethbowen) [inline test/lib/compile-nools-rules.spec.js:113]: There's no point mocking `/rules.nools.js` because you've just told it that it doesn't exist (fs.exists returns false on line 110).

What I think you need to do is mock `mocks.pack` to return 'http://foo.bar.com'.

File: src/lib/compile-nools-rules.js [hunk 2]
```diff
@@ -48,6 +48,9 @@ const compileDeclarativeFiles = async (projectDir, options) => {
   const baseEslintPath = path.join(__dirname, '../nools/.eslintrc');
   
   const code = await pack(projectDir, pathToDeclarativeLib, baseEslintPath, options);
+  if (code.match(/\/\/(.*)/g)) {
```
[the remark this answers is not in the document]
Comment (garethbowen) [reply src/lib/compile-nools-rules.js:51]: This implementation was chosen based on [this comment](https://github.com/medic/cht-core/issues/6506#issuecomment-648660236)

> I've never actually had to use a url in my tasks or targets.

Patching nools is not a very safe or sustainable approach, so we chose to look at cleaning up the input.

My preferred approach would be to escape URLs so they still output nicely but don't get picked up by the nools regex at all... ie:

1. pack will remove comments automatically (I think?), and comments in unpacked code shouldn't break nools, so we can ignore `// some comment` altogether
2. Detect and escape URLs specifically, eg: `://`

[the remark this answers is not in the document]
Comment (kennsippell) [reply src/lib/compile-nools-rules.js:51]: Generally, it sounds like parsing and editing javascript code using regex is what got the nools project into this problem. Perhaps it would be wise for us to avoid falling into the same trap?

[the remark this answers is not in the document]
Comment (garethbowen) [reply src/lib/compile-nools-rules.js:51]: > Liked this from Stefan

Yes, agreed. This PR needs more unit testing before merge.

> Perhaps it might be wise for us to avoid this pattern? Could we be falling into a similar trap that is causing this
problem in Nools?

Sure. The problem is the patching pattern has been a trap in the past too. One of the reasons I prefer this approach is I'd rather the patching was done in medic-conf and issues are found early and can be worked around, rather than doing it in the browser when it may be found too late.

Can you think of a solution that doesn't involve either updating the rule source code, or the library source code?

[the remark this answers is not in the document]
Comment (kennsippell) [reply src/lib/compile-nools-rules.js:51]: Sorry, no I can't think of a way :( We are between a rock and a hard place

I definitely see your point about the more adaptive release cycle for medic-conf. That sounds nice, since this is a sticky scenario. 

Some parting thoughts and then I'll let you guys do the hard work:

* Is it possible this approach would alter the behavior of code, or that code would behave different in unit tests vs after compilation.  What risks might that pose?
* How might this approach behave if run against code that is parsing a url, or code that contains a complex regex with slashes and colons?
* Which of the approaches reduces the most complexity and for which users. If the complexity of all solutions is the same, is it preferrable to put complexity on app builders, or on devs? 
* It seems we are currently "in the trap" of patching, but we have just recently escaped the trap of mutating build outputs
* Should this bug be left unfixed until 4.x when we just remove nools?

I'm sure you two will make the right choice. Gl

**Comment by njogz** [comment]
Hi @garethbowen. Please review when you can. Thanks.

**Review by garethbowen — CHANGES_REQUESTED** [review]
Thanks for dusting this off!

Firstly, whatever solution we go with, I think it needs more unit tests. For example, what happens if the code has a comment, does multiline work, any other edge cases you can think of.

Secondly, do you think this is the right solution? There's a discussion thread in the PR between Kenn and me about different approaches. It'd be good to get your 2c on that too!

File: src/lib/compile-nools-rules.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (njogz) [reply src/lib/compile-nools-rules.js:51]: I agree with @garethbowen on this as the better solution mainly because we can have a faster release cycle if another patch is needed.

[the remark this answers is not in the document]
Comment (njogz) [reply src/lib/compile-nools-rules.js:51]: @kennsippell we are unlikely to drop nools in 4.0 and this [issue](https://github.com/medic/cht-core/issues/5906) which touches on that is scheduled for 5.0. How do you feel about that given the multiple year timeline for 5.0?

[the remark this answers is not in the document]
Comment (kennsippell) [reply src/lib/compile-nools-rules.js:51]: my opinion is still that the code in this PR is not a step foward and has potential to block the re-use of more libraries. i'd encourage no-op and personally would wait to remove nools
